### PR TITLE
Start writing to offer and conditions tables when making and changing offers

### PR DESCRIPTION
**Pull request by benswannack**
## Context
We are migrating from the current offer structure, to a new model where we store conditions individually in their own table. This is the second step where we start writing to the new structure in parallel to the old structure.

Worked on with @richardpattinson 

## Changes proposed in this pull request
Write to Offer and OfferCondition models in:
MakeOffer
ChangeOffer
Support - change accepted offer conditions
Application choice factories

## Guidance to review
Deleting and recreating conditions is not ideal in terms of auditing. I think we will need to revisit this anyway when we start tracking condition statuses individually. For now this does the job of leaving us with conditions in the correct states at all times.

## Link to Trello card
https://trello.com/c/gFet4v13/3719-start-writing-to-offer-and-conditions-tables-when-making-and-changing-offers

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/write_offer.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    @conditions = conditions
+    @application_choice = application_choice
+  end
+
+  def call
+    offer = Offer.find_or_initialize_by(application_choice: @application_choice)
+    offer.conditions.destroy_all
+    @conditions.each do |condition|
+      OfferCondition.create(offer: offer, text: condition, status: condition_status)
+    end
```
Comment (benswannack) [inline app/services/write_offer.rb:12]: Something here wants me to include `:offer` in a query. Investigating...

File: app/services/write_offer.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class WriteOffer
+  def initialize(conditions:, application_choice:)
+    @conditions = conditions
+    @application_choice = application_choice
+  end
+
+  def call
+    offer = Offer.find_or_initialize_by(application_choice: @application_choice)
+    offer.conditions.destroy_all
```
Comment (benswannack) [inline app/services/write_offer.rb]: This leaves some orphaned audits. Should we delete them too?

File: app/services/write_offer.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class WriteOffer
+  def initialize(conditions:, application_choice:)
+    @conditions = conditions
+    @application_choice = application_choice
+  end
+
+  def call
+    offer = Offer.find_or_initialize_by(application_choice: @application_choice)
+    OfferCondition.where(offer: offer).delete_all
```
Comment (benswannack) [inline app/services/write_offer.rb:9]: This leaves some orphaned audits. Should we delete them too?

Comment (despo) [reply app/services/write_offer.rb:9]: I understand that this is easier to achieve, but I think we should only be deleting condition that no longer exist, and only creating new ones (or ones that have been reworded). That way all audit entries will also be relevant.

File: app/models/offer.rb [hunk 4]
```diff
@@ -1,4 +1,12 @@
 class Offer < ApplicationRecord
   belongs_to :application_choice
   has_many :conditions, class_name: 'OfferCondition'
+
+  def mark_all_conditions_as_met!
+    conditions.update(status: :met)
```
Comment (despo) [inline app/models/offer.rb:6]:
```suggestion
    conditions.update!(status: :met)
```

File: app/models/offer.rb [hunk 5]
```diff
@@ -1,4 +1,12 @@
...
+  end
+
+  def mark_all_conditions_as_not_met!
+    conditions.update(status: :unmet)
```
Comment (despo) [inline app/models/offer.rb:10]:
```suggestion
    conditions.update!(status: :unmet)
```

File: app/services/write_offer.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+class WriteOffer
```
Comment (despo) [inline app/services/write_offer.rb:1]: Can this be `UpdateOfferConditions` instead with `offer` and `conditions` as argument?

As offer is nested in application choice, we should be able to make any changes directly to the model as part of the Make and Change offer service, so I'm not sure if there is any value in having a dedicated service for it. 

However, as conditions do have an added complexity (specific status depending on application choice state, checking if they exist or not etc) it would make sense to have their own service.

File: spec/services/confirm_offer_conditions_spec.rb [hunk 7]
```diff
@@ -33,4 +33,18 @@
       expect(notifier).to have_received(:application_outcome_notification)
     end
   end
+
+  it 'sets the status of all the offer conditions to met' do
+    application_choice = create(:application_choice, :with_offer, status: :pending_conditions)
+    offer = Offer.find_by(application_choice: application_choice)
+
+    Timecop.freeze do
```
Comment (despo) [inline spec/services/confirm_offer_conditions_spec.rb:41]: Why is this needed?

File: app/services/write_offer.rb (hunk 3 above)
Comment (benswannack) [reply app/services/write_offer.rb:9]: Yeah, it would be nice to have it do that. It initially felt like more effort than it's worth though.

We can take the list of conditions passed in and match them to existing conditions. Then we are left with some new condition strings, and a list of unmapped existing OfferConditions. We'd have to randomly update the OfferConditions with the new condition strings, and add/delete OfferConditions to match the numbers.

All we are actually trying to achieve here is to keep the state of Offer & OfferCondition consistent with what's in `application_choice.offer` (more like a rolling backfill). When we split up the conditions to be managed individually, we will need to worry about this (in particular preserving condition statuses) and we'll be in a better position to make structural changes to support it.

What do you think?

File: app/models/offer.rb (hunk 4 above)
Comment (benswannack) [reply app/models/offer.rb:6]: `update!` is not a method on a collection. I'm doing this instead of `update_all` because `update_all` skips callbacks (and hence auditing)

File: spec/services/confirm_offer_conditions_spec.rb (hunk 7 above)
Comment (benswannack) [reply spec/services/confirm_offer_conditions_spec.rb:41]: Removed

File: app/models/offer.rb (hunk 4 above)
Comment (despo) [reply app/models/offer.rb:6]: I think that would be best, and then the method name would also correctly reflect that failure causes an exception to be thrown.

File: app/services/write_offer.rb (hunk 6 above)
Comment (benswannack) [reply app/services/write_offer.rb:1]: The idea of this service is to wrap up the logic of "I have a list of strings, please save them as conditions". Since this happens in a few places, this service feels sensible.

I imagine this service won't be required when we move to managing conditions individually though

File: app/services/write_offer.rb (hunk 3 above)
Comment (despo) [reply app/services/write_offer.rb:9]: As discussed after the standup, we can leave this like this for now.

File: app/services/write_offer.rb (hunk 6 above)
Comment (benswannack) [reply app/services/write_offer.rb]: We have renamed the service, but I think the behaviour of it is fine unless I've misunderstood you're point

File: app/services/write_offer.rb (hunk 3 above)
Comment (benswannack) [reply app/services/write_offer.rb]: We've disabled auditing on the OfferConditions model to avoid leaving orphaned audits. We will re-add this in the future

File: app/services/update_offer_conditions.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
+class UpdateOfferConditions
+  def initialize(conditions:, application_choice:)
```
Comment (despo) [inline app/services/update_offer_conditions.rb:2]: Nitpicking: Can we reverse the order of the arguments, as we rely on the application choice in order to do anything, so it makes sens to have it as they primary argument.

```suggestion
  def initialize(application_choice:, conditions:)
```

File: app/services/make_offer.rb [hunk 9]
```diff
@@ -25,6 +25,7 @@ def save!
 
           application_choice.current_course_option = course_option
           application_choice.offer = { 'conditions' => conditions }
+          UpdateOfferConditions.new(conditions: conditions, application_choice: application_choice).call
```
Comment (despo) [inline app/services/make_offer.rb:28]:
```suggestion
          UpdateOfferConditions.new(application_choice: application_choice, conditions: conditions).call
```

File: app/services/change_offer.rb [hunk 10]
```diff
@@ -23,6 +23,7 @@ def save!
 
           application_choice.current_course_option = course_option
           application_choice.offer = { 'conditions' => conditions }
+          UpdateOfferConditions.new(conditions: conditions, application_choice: application_choice).call
```
Comment (despo) [inline app/services/change_offer.rb:26]:
```suggestion
          UpdateOfferConditions.new(application_choice: application_choice, conditions: conditions).call
```

**Review by despo — APPROVED** [review]
Looks good. I just left some minor feedback about the order of the `UpdateOfferConditions` arguments.

**Review by cwrw — COMMENTED** [review]
are we also covering the deferred/reconfirm conditions flow?

File: app/models/offer.rb [hunk 11]
```diff
@@ -1,4 +1,12 @@
 class Offer < ApplicationRecord
   belongs_to :application_choice
   has_many :conditions, class_name: 'OfferCondition'
+
+  def mark_all_conditions_as_met!
+    conditions.each { |condition| condition.update!(status: :met) }
```
Comment (cwrw) [inline app/models/offer.rb:6]: can we avoid looping over each condition and update in bulk?

Comment (cwrw) [reply app/models/offer.rb:6]: saw previous comment nvm

**Comment by benswannack** [comment]
> are we also covering the deferred/reconfirm conditions flow?

That's a good point, we need to update condition statuses there too

File: app/models/offer.rb (hunk 11 above)
Comment (benswannack) [reply app/models/offer.rb:6]: We have 3 options here:
1. `conditions.update`
2. `conditions.update_all`
3. `conditions.each update!`

I originally didn't go with `update_all` because it skips callbacks and we had auditing. Now we've removed auditing we could go back to that.

The reason we went with this is to get it to throw if saving fails

Comment (cwrw) [reply app/models/offer.rb:10]: I think if we go with `update` it will ensure callbacks and bulk, but not failing on validations. However will we ever fail on validations here, I think probably unlikely at this stage, what do you think?

Comment (benswannack) [reply app/models/offer.rb:10]: Since we don't have validations, then I don't think we can fail at the moment. I'd be happy with just `update` here (which will fail if you give it something that's not in the enum)

File: app/services/update_offer_conditions.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
+class UpdateOfferConditions
+  def initialize(application_choice:, conditions:)
+    @application_choice = application_choice
+    @conditions = conditions
+  end
+
+  def call
+    offer = Offer.find_or_initialize_by(application_choice: @application_choice)
+    OfferCondition.where(offer: offer).delete_all
+    @conditions.each do |condition|
```
Comment (cwrw) [inline app/services/update_offer_conditions.rb:10]: quite a few queries happening here, is there something we can do with bulk `insert` methods?

File: app/models/offer.rb (hunk 11 above)
Comment (despo) [reply app/models/offer.rb:10]: Can we amend the method name then to exclude the bang suffix

File: app/services/update_offer_conditions.rb (hunk 12 above)
Comment (benswannack) [reply app/services/update_offer_conditions.rb:10]: `insert_all!` does let us do this, but we need to provide timestamps as it doesn't by default. It seems nicer to do this in one query though. Adding it now

File: app/services/conditions_not_met.rb [hunk 13]
```diff
@@ -13,6 +13,7 @@ def save
     audit(@auth.actor) do
       ApplicationStateChange.new(@application_choice).conditions_not_met!
       @application_choice.update!(conditions_not_met_at: Time.zone.now)
+      Offer.find_by(application_choice: @application_choice).mark_all_conditions_as_not_met
```
Comment (benswannack) [inline app/services/conditions_not_met.rb:16]: As discussed, we actually want to create an Offer here if it doesn't exist

**Review by cwrw — APPROVED** [review]
👍  Thanks for addressing comments looks great
